webhook/batch_url_notifier.go

```diff
@@ -0,0 +1,155 @@
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"github.com/livekit/protocol/auth"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+	"google.golang.org/protobuf/encoding/protojson"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const (
+	DefaultBatchSendInterval = 100 * time.Millisecond
+	DefaultMaxBatchSize      = 10000
+	DefaultRequestTimeout    = 10 * time.Second
+)
+
+type BatchURLNotifierParams struct {
+	Logger         logger.Logger
+	URL            string
+	Interval       time.Duration
+	RequestTimeout time.Duration
+	MaxSize        int
+	APIKey         string
+	APISecret      string
+}
+
+type BatchURLNotifier struct {
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	client     *http.Client
+	mu         sync.RWMutex
+	params     BatchURLNotifierParams
+	batch      []*livekit.WebhookEvent
+	dropped    int // it's operated inside a mutex scope so no need for atomic type
+}
+
+func NewBatchURLNotifier(ctx context.Context, params BatchURLNotifierParams) URLNotifier {
+	if params.Interval == 0 {
+		params.Interval = DefaultBatchSendInterval
+	}
+	if params.MaxSize == 0 {
+		params.MaxSize = DefaultMaxBatchSize
+	}
+	if params.RequestTimeout == 0 {
+		params.RequestTimeout = DefaultRequestTimeout
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	notifier := &BatchURLNotifier{
+		ctx:        ctx,
+		cancelFunc: cancel,
+		params:     params,
+		client:     http.DefaultClient,
+	}
+	notifier.client.Timeout = params.RequestTimeout
+
+	go notifier.runner()
+
+	return notifier
+}
+
+func (b *BatchURLNotifier) runner() {
+	ticker := time.NewTicker(b.params.Interval)
+	for {
+		select {
+		case <-ticker.C:
+			b.mu.Lock()
+			b.sendBatch()
+			b.mu.Unlock()
+		case <-b.ctx.Done():
+			return
+		}
+	}
+}
+
+func (b *BatchURLNotifier) sendBatch() {
+	raw := &livekit.BatchedWebhookEvents{
+		Events:     b.batch,
+		NumDropped: int32(b.dropped),
+		DequeuedAt: time.Now().Unix(),
+	}
+	defer func() {
+		b.batch = nil
+	}()
+	b.dropped = 0
+
+	encoded, err := protojson.Marshal(raw)
+	if err != nil {
+		b.dropped += len(b.batch)
```

Comment:
maybe add a warning log for this error? Otherwise, it's hard to identify something is wrong with marshaling.